xdsl/transforms/jax_use_donated_arguments.py

```diff
@@ -0,0 +1,144 @@
+import itertools
+from collections.abc import Sequence
+from dataclasses import dataclass
+
+from xdsl.context import MLContext
+from xdsl.dialects import builtin
+from xdsl.dialects.bufferization import MaterializeInDestinationOp
+from xdsl.dialects.builtin import FunctionType, TensorType
+from xdsl.dialects.func import FuncOp, ReturnOp
+from xdsl.ir import BlockArgument, Operation, SSAValue
+from xdsl.irdl import VarOperand
+from xdsl.passes import ModulePass
+from xdsl.pattern_rewriter import (
+    PatternRewriter,
+    PatternRewriteWalker,
+    RewritePattern,
+    op_type_rewrite_pattern,
+)
+
+
+def map_outputs_to_inputs(
+    donated_inputs: Sequence[BlockArgument], outputs: VarOperand
+) -> dict[int, int]:
+    """
+    Find suitable donated buffers for each of returned variables.
+    Each buffer can be used only once.
+    Types of the buffer and the variable should match.
+    """
+    used_inputs_idx: set[int] = set()
+    output_input_mapping: dict[int, int] = {}
+
+    for outputs_idx, output in enumerate(outputs):
+        if isinstance(output.owner, MaterializeInDestinationOp) and isinstance(
+            output.owner.dest, BlockArgument
+        ):
+            # the return value is already bufferized in a donated buffer
+            continue
+        for input_idx, arg in enumerate(donated_inputs):
+            if arg.type == output.type and input_idx not in used_inputs_idx:
+                output_input_mapping[outputs_idx] = input_idx
+                used_inputs_idx.add(input_idx)
+                break
+
+    return output_input_mapping
+
+
+def construct_new_output_list(
+    donated_inputs: Sequence[BlockArgument],
+    outputs: VarOperand,
+    output_input_mapping: dict[int, int],
+) -> tuple[Sequence[SSAValue], list[Operation]]:
+    """
+    Create new SSA values of buffers with the needed content, they will be used in the ReturnOp.
+    Also create operations to associate buffers with corresponding return values.
+    """
+    new_outputs: Sequence[SSAValue] = [o for o in outputs]
+    materialize_ops: list[Operation] = []
+
+    for output_idx, input_idx in sorted(output_input_mapping.items()):
+        materialize_ops.append(
+            MaterializeInDestinationOp(
+                operands=[outputs[output_idx], donated_inputs[input_idx]],
+                result_types=[outputs[output_idx].type],
+            )
+        )
+        new_outputs[output_idx] = materialize_ops[-1].results[0]
+
+    return new_outputs, materialize_ops
```

Comment:
If the mapping is changed to be of type `dict[SSAValue, BlockArgument]`, then each of the results of this helper can be written in a one-liner, and probably better inlined then. The rule of thumb for me is that a helper is useful to isolate some mutation or tricky state, and if we can eliminate the mutation or tricky state entirely then that seems like a good move. Sorry to yo-yo on this, sometimes the complexity only goes away after some iteration!